### PR TITLE
Allow users to define ctl_connection_settings and the option to fail when table information is unavailable

### LINKED ISSUE #56: Strange error in event_row.py
Hi,

Nice library. 

Here is a problem that has started happening, and I don't know why. Here is the Python error:

 File "/usr/lib/python2.6/site-packages/pymysqlreplication/binlogstream.py", line 146, in fetchone
    self.**use_checksum)
  File "/usr/lib/python2.6/site-packages/pymysqlreplication/packet.py", line 93, in __init**
    ctl_connection)
  File "/usr/lib/python2.6/site-packages/pymysqlreplication/row_event.py", line 491, in **init**
    column_schema = self.column_schemas[i]
IndexError: tuple index out of range

I think this is due to some MySQL internal query that pymysqlreplication has trouble with. The EventType that appears to trigger the failure is a QueryEvent, but the event is an internal MySQL event. Here some events that occur just before the failure. You can see that these are internal MySQL queries:

=== QueryEvent ===
Date: 2013-12-11T16:31:15
Log position: 30510
Event size: 80
Read bytes: 80
Schema: mysql
Execution time: 0
Query: TRUNCATE TABLE time_zone_transition
=== QueryEvent ===
Date: 2013-12-11T16:31:15
Log position: 30614
Event size: 85
Read bytes: 85
Schema: mysql
Execution time: 0
Query: TRUNCATE TABLE time_zone_transition_type
=== QueryEvent ===
Date: 2013-12-11T16:31:15
Log position: 30683
Event size: 50
Read bytes: 50
Schema: mysql
Execution time: 0
Query: BEGIN

^^ error occurs immediately after this

You can see that the date is 2013.12.11.

I looked at the source code where the error occurs:

485        for i in range(0, len(column_types)):
486            column_type = column_types[i]
487            column_schema = self.column_schemas[i]
488            col = Column(byte2int(column_type), column_schema, from_packet)
489            self.columns.append(col)

The problem is that self.column_schemas is None, so index fails. 

Any ideas on this? Thanks.

### PR DESCRIPTION
**Pull request by kdparker**
Resolves #56 
To allow users to create their own secondary database that tracks schema changes if they want to ensure they will always have table_metadata (in case of a dropped table before a message can be picked up). It also adds an (optional) parameter that makes the BinLogStreamReader raise an exception if there is an issue getting schema information for a query. This allows the library to stay low-level while giving users the option of alleviating this issue.

### CODE REVIEW

**Comment by kdparker** [comment]
By the way, if you want me to squash these commits, let me know.

File: .gitignore [hunk 1]
```diff
@@ -37,6 +37,9 @@ _build
 *.xml
 *.iml
 
+# Temp vim .swp s
+*.swp
```
Comment (baloo) [inline .gitignore]: Pro tip: this should go in your global ignore file instead. Defaults to: $HOME/.config/git/ignore

This is not blocking for merge though.

File: pymysqlreplication/binlogstream.py [hunk 2]
```diff
@@ -150,20 +152,28 @@ def __init__(self, connection_settings, server_id, resume_stream=False,
...
             report_slave: Report slave in SHOW SLAVE HOSTS.
             slave_uuid: Report slave_uuid in SHOW SLAVE HOSTS.
+            fail_on_table_metadata_unavailable: Should raise exception if we can't get
+                                                table information on row_events
         """
+
         self.__connection_settings = connection_settings
-        self.__connection_settings["charset"] = "utf8"
+        if not connection_settings.get("charset"):
+            self.__connection_settings["charset"] = "utf8"
```
Comment (baloo) [inline pymysqlreplication/binlogstream.py]:
``` python
self.__connection_settings.setdefault("charset", "utf8")
```

**Comment by noplay** [comment]
I agree the change sound good. If you can add a test for ctl_connection_settings it will be perfect

**Comment by kdparker** [comment]
Tests are done and working now (sorry for commit spam earlier, they're all squashed now)

**Comment by baloo** [comment]
\o/ I started looking at implementing the tests myself, and I have to admit I became quite lazy when looking at the travis mess. Thank you very much for your work. All looks good to me.

**Pull request merged**